crates/sui-framework/sources/transfer.move

```diff
@@ -14,6 +14,7 @@ module sui::transfer {
 
     /// Transfer ownership of `obj` to another object `owner`.
     public fun transfer_to_object<T: key, R: key>(obj: T, owner: &mut R) {
+        object::increment_child_count(owner);
```

Comment:
Where do we decrement the child count of its current parent?
Also, would it work if you are transferring a child object to its current parent (i.e. a noop transfer).

Given all the dances we have to do in the adapter anyway, I am starting to wonder whether we should put the count inside Move struct or should we keep it only in the rust struct.

Reply:
That is a very good question on the no-op transfer. I'll add a test for it

But I think the incrementing logic is just a tad more complicated to do outside of Move, but maybe I'm wrong. It just felt natural and easy to do it here.
Like its 1 line of Move, versus a whole complicated system in the adapter.
Particularly what is difficult outside of Move is switching from None to Some(1) as is done here. Because you have to shift around all the content bytes in the `MoveObject` 

Nonetheless, do you see any benefit from doing the increments in the adapter?

Reply:
> Where do we decrement the child count of its current parent?

Answering this separately, 
It is complicated but:
- In the adapter we keep track of all of the decrements
- Then at the end of the successful transaction, we apply them
- Then we check the valid rules around deletions/freezing and still having children

I'm actually going to change this code a bit, and you will see all of it in the next fixup